### PR TITLE
Linux texture encoding

### PR DESCRIPTION
**Pull request by kkartaltepe**
### Description

This enables the texture encoding path for opengl, my target is linux but I think we probably can/should support these on  mac/opengl so I did not restrict these to just windows+linux.

It is largely a rebase of some of commits in  #4974 that were specific just to enabling the texture_encode path on opengl.

It also adds a profiler root for the gpu encoder thread which is helpful when bringing up a new encoder.

### Motivation and Context

I was looking into texture encoding on QSV and the existing work in #4974 worked well in my implementation so it seems worth pulling into its own PR.

Considering I am working with vaapi/mesa and  #4974 is nvenc, I think this should be a workable solution for all platforms. But it would be good to have more discussion on whether this is the best way to pass textures. 

At least one consideration was whether we should pass DMABUFs directly, but in both my implementation and NVENC we need to use the opengl context to blit data into the encoder surfaces so passing a DMABUF just results in extra work to re-import it into opengl.

### How Has This Been Tested?

Tested in the implementation of QSV texture encoding in https://github.com/kkartaltepe/obs-studio/tree/qsv-linux-pt3 which "Works on my machine" at least.

### Types of changes

 - New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by kkartaltepe** [comment]
Note this is draft because I think there probably needs more discussion on if this is the best way to solve the texture encoding on opengl problem, but otherwise the code is done.

**Comment by norihiro** [comment]
I'd like to suggest revising the commit message of 3ec5a3a82 like below.
```plain
libobs,libobs-opengl: Enable GPU encoding for OpenGL

Enable all of the previously Windows only paths for OpenGL backends that
support encode_texture2
```

**Comment by kkartaltepe** [comment]
> I'd like to suggest revising the commit message of [3ec5a3a](https://github.com/obsproject/obs-studio/commit/3ec5a3a826c4a5548abf1a096d7a9bcf575b5f28) like below.
> 
> ```
> libobs,libobs-opengl: Enable GPU encoding for OpenGL
> 
> Enable all of the previously Windows only paths for OpenGL backends that
> support encode_texture2
> ```

Thanks, done and taking it out of WIP based on some discussion in discord today. As it seems most implementations seem fine with this structure and we can work out any more changes on the PR.

**Comment by kkartaltepe** [comment]
Rebased over the frameskip PR, we just happened to touch the same block to add variables. So this is good to go again.

**Comment by kkartaltepe** [comment]
Rebased after ROI conflicts and updated comments.

**Comment by kkartaltepe** [comment]
The qsv impl using this is also rebased and tested now.

File: libobs/obs-video-gpu-encode.c
Comment (derrod) [inline libobs/obs-video-gpu-encode.c:147]: What's this about?

Comment (kkartaltepe) [reply libobs/obs-video-gpu-encode.c:147]: The most difficult code to read is always your past self... I believe msvc blew up when trying to zero initialize this struct, so this added a byte-array of sufficient size. Lets remove it and see the error in CI.

Comment (derrod) [reply libobs/obs-video-gpu-encode.c:147]: It seems like even if MSVC hates it that it would be saner to just `memset` the struct rather than... this.

Comment (kkartaltepe) [reply libobs/obs-video-gpu-encode.c]: Actually i think this was left over from when we used a VLA instead of a sized array for gs_texture pointers. At least its working with msvc so monstrosity removed. Thanks.

File: libobs/obs-encoder.h
Comment (derrod) [inline libobs/obs-encoder.h:312]: This is a leftover from the ARGB PR and can be removed for now as it's unused. We shall revisit ARGB texture support later but right now this is unused.

File: libobs/obs-encoder.h
Comment (derrod) [inline libobs/obs-encoder.h:131]: This appears to be unused in all existing implementations and can probably be removed. Even if we add ARGB support we probably don't need this as all textures passed on an encoder must have the same format which should be negotiated on encoder init.

File: libobs/obs-video-gpu-encode.c
Comment (derrod) [inline libobs/obs-video-gpu-encode.c:141]: As noted above we don't really need the video info in the texture metadata so this can be removed.

File: libobs/obs-encoder.h
Comment (derrod) [inline libobs/obs-encoder.h:135]: Also not really sure about this, Why does it need to be NULL terminates? The currently in use texture format should determine the number of textures in use. And at most this would then be 4 (Y + U + V + Alpha).

**Review by derrod — CHANGES_REQUESTED** [review]
A couple leftovers from the old ARGB and texture encoding PRs that I think can be removed until we actually want to add support for more formats.

File: libobs/obs-encoder.h
Comment (kkartaltepe) [reply libobs/obs-encoder.h:135]: The 5th element was an earlier discussion to just be safe and include a terminal, i dont feel strongly.

As this is OBS's api we can decide to always use multiple textures, one per plane. In theory (in reality on vulkan) some platforms only do planar as multiple views on a single alloc, but we can cross that bridge when we get to it.

**Comment by kkartaltepe** [comment]
> A couple leftovers from the old ARGB and texture encoding PRs that I think can be removed until we actually want to add support for more formats.

Good catches, updated.

**Review by derrod — APPROVED** [review]
Should be fine as it is now, we have some time befire it'll make it into a stable release so revision should be no problem.

File: libobs/obs-encoder.c
Comment (tt2468) [inline libobs/obs-encoder.c:712]: What's happening here? I don't think it's a good idea to be changing any data with a `get` method.

File: libobs/obs-encoder.c
Comment (tt2468) [inline libobs/obs-encoder.c:697]: Is this necessary, or just a helpful util?

File: libobs/obs-encoder.h
Comment (tt2468) [inline libobs/obs-encoder.h:298]: If we're introducing an encoder_texture struct, it seems like `pts`, `lock_key`, and `next_key` should be values of that struct.

**Review by tt2468 — CHANGES_REQUESTED** [review]
I've tested this with Rodney's PR on Ubuntu 22.04 with an RTX A4000 GPU, running X11. It works well, and I tried a number of different codepaths like color space and other formats, and everything appears to work as desired. Great to see PCIe bus bandwidth staying in the 10's of MB/s while streaming...

File: libobs/obs-encoder.c
Comment (derrod) [reply libobs/obs-encoder.c:712]: This is just a copy of the existing `get_video_info`, also a leftover from the original ARGB stuff and I think it could be removed as well.

https://github.com/obsproject/obs-studio/blob/fbf6c8d243714b5d5409826f69c633c627d3535c/libobs/obs-encoder.c#L168

Comment (kkartaltepe) [reply libobs/obs-encoder.c:712]: It seemed useful and I assumed it was used in the ARGB stuff, but I also cannot find it used in the PR (and the old get_video_info is still leftover). Given its not actually used Ill remove it and revert to the internal.

That said do we still need that set on this path?

Comment (derrod) [reply libobs/obs-encoder.c:712]: The video info feature is to allow the encoder to specify the input format it wants, in theory I suppose this could include a different size. I'd leave the existing internal function alone for now.

File: libobs/obs-encoder.h
Comment (derrod) [reply libobs/obs-encoder.h:298]: pts sure, the next_key is technically an output variable so making that part of the struct feeels kind of ugly.

Comment (kkartaltepe) [reply libobs/obs-encoder.h:298]: im not sure i agree with any of them, its just that we dont share/reuse textures that all of these happen to have the same lifetimes.

Comment (derrod) [reply libobs/obs-encoder.h:298]: I see. What I have been wondering if perhaps we can move at least the pts to a struct similar to `encoder_frame`, the motivation here is that for the multi-encoder streaming Twitch is doing and #10096 we'll need a way to force IDR frames. 
For non-texture encoders we can simply extend the existing struct without breaking ABI, but for GPU encoders that isn't possible. So perhaps we can make this API more future-looking by adding a means to add more information later without requiring a new ABI?

I realise this is getting into bikeshedding territory, I won't block this PR on that, just wanted to hear your opinion.

**Review by tt2468 — APPROVED** [review]

**Review by RytoEX — APPROVED** [review]
Seems fine. Want to land this to unblock other ongoing work to give that work some lead time. Reviews here plus off-thread discussion indicate that this is fine to land.

**Pull request merged**
